src/main/java/com/mongodb/spark/sql/connector/config/AbstractMongoConfig.java

```diff
@@ -0,0 +1,328 @@
+  public MongoNamespace getNamespace() {
+    return new MongoNamespace(getDatabaseName(), getCollectionName());
+  }
+
+  /** @return the connection string */
+  @Override
+  public ConnectionString getConnectionString() {
+    return new ConnectionString(getOrDefault(CONNECTION_STRING_CONFIG, CONNECTION_STRING_DEFAULT));
+  }
+
+  @Override
+  public Map<String, String> getOptions() {
+    return options;
+  }
+
+  /** @return the {@link MongoClientFactory} for this configuration. */
+  MongoClientFactory getMongoClientFactory() {
+    if (mongoClientFactory == null) {
+      String mongoClientFactoryName =
+          getOptions().getOrDefault(CLIENT_FACTORY_CONFIG, CLIENT_FACTORY_DEFAULT);
+      mongoClientFactory =
+          ClassHelper.createInstance(
+              CLIENT_FACTORY_CONFIG, mongoClientFactoryName, MongoClientFactory.class, this);
+    }
+    return mongoClientFactory;
+  }
+
+  /**
+   * Returns a MongoClient
+   *
+   * <p>Once the {@link MongoClient} is no longer required, it MUST be closed by calling {@code
+   * mongoClient.close()}.
+   *
+   * @return the MongoClient from the cache or create a new one using the {@code
+   *     MongoClientFactory}.
+   */
+  public MongoClient getMongoClient() {
+    return LazyMongoClientCache.getMongoClient(getMongoClientFactory());
+  }
+
+  /**
+   * Loans a {@link MongoClient} to the user.
+   *
+   * @param function the function that is passed the {@code MongoClient}
+   * @param <T> The return type
+   * @return the result of the function
+   */
+  public <T> T withClient(final Function<MongoClient, T> function) {
+    try (MongoClient client = getMongoClient()) {
+      return function.apply(client);
+    }
+  }
+
+  /**
+   * Loans a {@link MongoClient} to the user, does not return a result.
+   *
+   * @param consumer the consumer of the {@code MongoClient}
+   */
+  public void doWithClient(final Consumer<MongoClient> consumer) {
+    withClient(
+        client -> {
+          consumer.accept(client);
+          return null;
+        });
+  }
+
+  /**
+   * Loans a {@link MongoCollection} to the user, does not return a result.
+   *
+   * @param consumer the consumer of the {@code MongoCollection<BsonDocument>}
+   */
+  public void doWithCollection(final Consumer<MongoCollection<BsonDocument>> consumer) {
+    doWithClient(
+        c ->
+            consumer.accept(
+                c.getDatabase(getDatabaseName())
+                    .getCollection(getCollectionName(), BsonDocument.class)));
+  }
+
+  @Override
+  public String toString() {
+    return "MongoConfig{" + "options=" + options + ", usageMode=" + usageMode + '}';
+  }
+
+  @TestOnly
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    final AbstractMongoConfig that = (AbstractMongoConfig) o;
+    return Objects.equals(options, that.options) && usageMode == that.usageMode;
```

Comment:
We only have an equals method for testing purposes and the originals can contain values that aren't relevant for the current usage mode. I think we could remove the equals method as part of SPARK-313 and only compare values from the getters.

Reply:
> originals can contain values that aren't relevant for the current usage mode

Even if the current implementation does not violate the `equals` requirements, it definitely violates the expectations, especially given that the method is not documented to warn users. The expectation I refer to is that `equal` objects behave equally, and it is not the case for the current implementation.

> We only have an equals method for testing purposes

While it may be true that the mongo-spark codebase uses this method only in tests, there is no way to technically prevent users from using the method, and since it's not even our method, but a method coming from `Object`, the default expectations that a user has is that it can be used, so one may not even look at the docs to see that we are trying to forbid using it. Moreover, we are simply not allowed to forbid to use a method that comes from `Object.equals` because this breaks its contract. Therefore, we must not mark the method with `@TestOnly`.

Reply:
I disagree here - equality checks are only used in testing and have no practical purpose else where.

Leaving as is.